lahja/endpoint.py

```diff
@@ -190,61 +200,23 @@ def __init__(self, conn: Connection, new_msg_func: Callable[[Broadcast], None])
         async with self._notify_lock:
             # The lock ensures only one coroutine can notify this endpoint at any one time
             # and that no replies are accidentally received by the wrong coroutines.
-            async with self._received_response:
+            self.logger.info('NOTIFY SUBSCRIPTIONS UPDATED: %s', subscriptions)
+            async with self._received_ack:
                 try:
                     await self.conn.send_message(SubscriptionsUpdated(subscriptions, block))
                 except RemoteDisconnected:
                     return
                 if block:
-                    await self._received_response.wait()
-
-
-class OutboundConnection:
-    """
-    The local Endpoint might have several ``OutboundConnection``s, each of them represents a
-    remote ``Endpoint`` which has been connected to. The remote endpoint occasionally sends
-    special message "backwards" to the local endpoint that connected to it.
-
-    Those messages (``SubscriptionsUpdated``) specify which kinds of messages the remote
-    Endpoint is subscribed to. No other message types are allowed to flow "backwards" from
-    an outbound connection and otherwise are dropped.
-    """
-    def __init__(self, name: str, conn: Connection) -> None:
-        self.conn = conn
-        self.name = name
-        self.subscribed_messages: Set[Type[BaseEvent]] = set()
-
-        self.logger = logging.getLogger('lahja.endpoint.OutboundConnection')
-        self._received_subscription = asyncio.Condition()
-
-    async def run(self) -> None:
-        while True:
-            try:
-                message = await self.conn.read_message()
-            except RemoteDisconnected:
-                return
-
-            if not isinstance(message, SubscriptionsUpdated):
-                self.logger.error(
-                    f'Endpoint {self.name} sent back an unexpected message: {type(message)}'
-                )
-                return
-
-            self.subscribed_messages = message.subscriptions
-            async with self._received_subscription:
-                self._received_subscription.notify_all()
-            if message.response_expected:
-                await self.send_message(SubscriptionsAck())
+                    await self._received_ack.wait()
 
     def can_send_item(self, item: BaseEvent, config: Optional[BroadcastConfig]) -> bool:
-        is_response = config is not None and config.filter_event_id is not None
-        passes_config = config is None or config.allowed_to_receive(self.name)
-        passes_filter = type(item) in self.subscribed_messages
-
-        if not passes_config:
-            return False
+        if config is not None:
+            if config.filter_event_id is not None:
+                return True
+            elif not config.allowed_to_receive(self.name):
+                return False
 
-        return is_response or passes_filter
+        return type(item) in self.subscribed_messages
```

Comment:
Want to profile this as well, intended to optimize this function for early returns and not pre-computing un-needed information in those cases.